File: tests/java/tests-java-gradle-junit5-allure3-selenium/src/test/java/pages/LoginPage.java
package pages;

import helpers.Ui;
import io.qameta.allure.Step;
import org.openqa.selenium.WebElement;

public class LoginPage extends BasePage<LoginPage> {

    @Step("Open login page")
    public LoginPage openPage() {
        Ui.open("/login");
        return shouldBeOpen();
    }

    @Step("Click Register link under the login form")
    public RegisterPage clickRegisterLink() {
        Ui.click("register-link");
        return new RegisterPage();
    }

    @Step("Fill and submit form")
    public HomePage fillAndSubmitForm(String username, String password) {
        typeUsername(username);
        typePassword(password);
        return submit();
    }

    @Step("Type username: {username}")
    public LoginPage typeUsername(String username) {
        Ui.setValue("login-input", username);
        return this;
    }

    @Step("Type password")
    public LoginPage typePassword(String password) {
        Ui.setValue("password-input", password);
        return this;
    }

    @Step("Submit login form")
    public HomePage submit() {
        Ui.click("submit-button");
        return new HomePage();
    }

    @Step("Submit login form expecting validation error")
    public LoginPage submitExpectingError() {
        Ui.click("submit-button");
        Ui.shouldBeVisible("error-message");
        return this;
    }

    @Override
    @Step("Verify login page is open")
    public LoginPage shouldBeOpen() {
        Ui.shouldBeVisible("login-form");
        return this;
    }

    @Step("Verify login form is mounted")
    public LoginPage shouldShowLoginForm() {
        Ui.shouldBeVisible("login-form-title");
        Ui.shouldBeVisible("login-input");
        Ui.shouldBeVisible("password-input");
        Ui.shouldBeVisible("submit-button");
        return this;
    }

    @Step("Login form panel is visible")
    public WebElement loginFormPanel() {
        return Ui.el("login-form");
    }

    @Step("Verify form title message: {message}")
    public LoginPage shouldHaveFormTitle(String message) {
        Ui.shouldHaveText("login-form-title", message);
        return this;
    }

    @Step("Verify error message: {message}")
    public LoginPage shouldHaveErrorMessage(String message) {
        Ui.shouldHaveText("error-message", message);
        return this;
    }
}

File: tests/csharp/tests-csharp-nunit-restsharp-selenium/Pages/LoginPage.cs
using Allure.NUnit.Attributes;
using Helpers;
using OpenQA.Selenium;

namespace Pages;

public sealed class LoginPage : BasePage<LoginPage>
{
    [AllureStep("Open login page")]
    public LoginPage OpenPage()
    {
        Ui.Open("/login");
        return ShouldBeOpen();
    }

    [AllureStep("Click Register link under the login form")]
    public RegisterPage ClickRegisterLink()
    {
        Ui.Click("register-link");
        return new RegisterPage();
    }

    [AllureStep("Fill and submit form")]
    public HomePage FillAndSubmitForm(string username, string password)
    {
        TypeUsername(username);
        TypePassword(password);
        return Submit();
    }

    [AllureStep("Type username: {username}")]
    public LoginPage TypeUsername(string username)
    {
        Ui.SetValue("login-input", username);
        return this;
    }

    [AllureStep("Type password")]
    public LoginPage TypePassword(string password)
    {
        Ui.SetValue("password-input", password);
        return this;
    }

    [AllureStep("Submit login form")]
    public HomePage Submit()
    {
        Ui.Click("submit-button");
        return new HomePage();
    }

    [AllureStep("Submit login form expecting validation error")]
    public LoginPage SubmitExpectingError()
    {
        Ui.Click("submit-button");
        Ui.ShouldBeVisible("error-message");
        return this;
    }

    [AllureStep("Verify login page is open")]
    public override LoginPage ShouldBeOpen()
    {
        Ui.ShouldBeVisible("login-form");
        return this;
    }

    [AllureStep("Verify login form is mounted")]
    public LoginPage ShouldShowLoginForm()
    {
        Ui.ShouldBeVisible("login-form-title");
        Ui.ShouldBeVisible("login-input");
        Ui.ShouldBeVisible("password-input");
        Ui.ShouldBeVisible("submit-button");
        return this;
    }

    [AllureStep("Login form panel is visible")]
    public IWebElement LoginFormPanel() => Ui.El("login-form");

    [AllureStep("Verify form title message: {message}")]
    public LoginPage ShouldHaveFormTitle(string message)
    {
        Ui.ShouldHaveText("login-form-title", message);
        return this;
    }

    [AllureStep("Verify error message: {message}")]
    public LoginPage ShouldHaveErrorMessage(string message)
    {
        Ui.ShouldHaveText("error-message", message);
        return this;
    }
}
